[DXIL][SPIRV] Lower `WaveActiveCountBits` intrinsic

inbelic opened the pull request:
```
  - add codegen for llvm builtin to spirv/directx intrinsic in CGBuiltin.cpp
  - add lowering of spirv intrinsic to spirv backend in SPIRVInstructionSelector.cpp
  - add lowering of directx intrinsic to dxil op in DXIL.td

  - add test cases to illustrate passes
  - add test case for semantic analysis
  ```
  
  Resolves #80176

inbelic committed:
[DXIL][SPIRV] Lower WaveActiveCountBits intrinsic

  - add codegen for llvm builtin to spirv/directx intrinsic
    in CGBuiltin.cpp
  - add lowering of spirv intrinsic to spirv backend in
    SPIRVInstructionSelector.cpp
  - add lowering of directx intrinsic to dxil op in DXIL.td

  - add test cases to illustrate passes
  - add test case for semantic analysis

Commit:
clang-format

inbelic commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1762,6 +1765,36 @@ bool SPIRVInstructionSelector::selectSign(Register ResVReg,
+          .addDef(ResVReg)
+          .addUse(GR.getSPIRVTypeID(ResType))
+          .addUse(GR.getOrCreateConstInt(SPIRV::Scope::Subgroup, I, IntTy, TII))
+          .addImm(0)
```
Will update to use the `SPIRV::OperationGroup` enum

pow2clk replied:
Do I understand that you'll do this before submitting this PR?

inbelic replied:
Yep, have it added for when the other review comments are addressed.

inbelic commented:
Failing testcase is unrelated. Will need to rebase on review changes.

pow2clk commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1762,6 +1765,36 @@ bool SPIRVInstructionSelector::selectSign(Register ResVReg,
+          .addDef(BallotReg)
+          .addUse(GR.getSPIRVTypeID(BallotType))
+          .addUse(GR.getOrCreateConstInt(SPIRV::Scope::Subgroup, I, IntTy, TII))
+          .addUse(I.getOperand(2).getReg());
```
You're missing a `constrainAllUses` call on the end here. `AddUse` returns a reference to `MachineInstrBuilder`, which I'm not really sure how gets converted to a bool, but I don't want to rely on it.

pow2clk commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -820,3 +820,12 @@ def WaveGetLaneIndex : DXILOp<111, waveGetLaneIndex> {
   let attributes = [Attributes<DXIL1_0, [ReadNone]>];
 }
+
+def WaveAllBitCount : DXILOp<135, waveAllOp> {
```
Initially, I thought this couldn't be right since the latter name `waveAllOp` mismatches the former `WaveAllBitCount` in more than just capitalization, but though DXIL.rst makes no mention of waveAllOp, that is the name of the actual DXIL op. Weird, but consistent.

pow2clk reviewed (commented):
Looks good! The only change I think is essential is the missing `constrainAllUses`

inbelic commented on clang/test/CodeGenHLSL/builtins/WaveActiveCountBits.hlsl:
```diff
@@ -0,0 +1,22 @@
+// CHECK-LABEL: test_bool
+int test_bool(bool expr) {
+  // CHECK-SPIRV: %[[#entry_tok:]] = call token @llvm.experimental.convergence.entry()
+  // CHECK-SPIRV:  %[[RET:.*]] = call spir_func i32 @llvm.spv.wave.active.countbits(i1 %{{.*}}) [ "convergencectrl"(token %[[#entry_tok]]) ]
```
I guess when the change to add the convergence token was first introduced it was important to check that the convergence token was correctly generated for the appropriate spirv operations. But now it is probably quite redundant to add that on each test, and checking for the convergent attribute is sufficient.

Will update to use `-DTARGET` as suggested.

inbelic committed:
review comments:

  - add constrainAllUses to first spirv op
  - update testcase for ease of reading
  - use enum instead of int equivalent for documentation

inbelic commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -820,3 +820,12 @@ def WaveGetLaneIndex : DXILOp<111, waveGetLaneIndex> {
+  let arguments = [Int1Ty];
+  let result = Int32Ty;
+  let stages = [Stages<DXIL1_0, [all_stages]>];
+  let attributes = [Attributes<DXIL1_0, [ReadNone]>];
```
@pow2clk Just want to ping to double check these attributes are okay. `ReadNone` is not in `hctdb.py` but it would seem to be applicable. Should `IsWave` be added now or will we add them all at once in a separate commit?

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1762,6 +1765,37 @@ bool SPIRVInstructionSelector::selectSign(Register ResVReg,
+  assert(I.getOperand(2).isReg());
+  MachineBasicBlock &BB = *I.getParent();
+
+  Register BallotReg = MRI->createVirtualRegister(&SPIRV::IDRegClass);
```
Is SPIRV::IDRegClass just a register class for all types? Because there are others like SPIRV::vIDRegClass, which is for vectors.

inbelic replied:
Good catch. Changed to using `GR.getRegClass(BallotType)`.

inbelic committed:
review comments:

- get the proper register class
- use result and instead of or

inbelic commented:
Rebased to resolve conflicts.

inbelic committed:
Merge branch 'main' into inbelic/wave-active-bitcount

inbelic committed:
Merge branch 'main' into inbelic/wave-active-bitcount

inbelic committed:
Merge branch 'main' into inbelic/wave-active-bitcount
